File: Day7_Number_of_Substrings_Containing_All_Three_Characters/NumberOfSubstrings.java
class Solution {
    public int numberOfStringsLessOrEqualThreeChar(String s,int k){
        HashMap<Character,Integer> m=new HashMap<>();
        int p1=0;
        int p2=0;
        int ans=0;

        while(p1<s.length()){
            if( m.containsKey(s.charAt(p1))){
                m.put(s.charAt(p1),m.get(s.charAt(p1))+1);
            }
            else{
                m.put(s.charAt(p1),1);
            }
            if(m.size()<=k){
                ans+=(p1-p2+1);
            }
            else{
                while(m.size()>k){
                    m.put(s.charAt(p2),m.get(s.charAt(p2))-1);
                    if (m.get(s.charAt(p2))==0){
                        m.remove(s.charAt(p2));
                    }
                    p2+=1;
                }
                ans+=(p1-p2+1);
            }
            p1+=1;
        }
        return ans;


    }
    public int numberOfSubstrings(String s) {
        return numberOfStringsLessOrEqualThreeChar(s,3)-numberOfStringsLessOrEqualThreeChar(s,2);
        
    }
}

File: Day7_Number_of_Substrings_Containing_All_Three_Characters/NumberOfSubstrings.py

class Solution(object):
    def numberOfStringsLessOrEqualThreeChar(self,s,k):
        d=dict()
        p1=0
        p2=0
        ans=0
        while p1<len(s):
            if s[p1] in d:
                d[s[p1]]+=1
            else:
                d[s[p1]]=1
            if len(d)<=k:
                ans+=(p1-p2+1)
            else:
                while len(d)>k:
                    d[s[p2]]-=1
                    if d[s[p2]]==0:
                        del d[s[p2]]
                    p2+=1
                ans+=(p1-p2+1)
            p1+=1
        return ans
    def numberOfSubstrings(self, s):
        """
        :type s: str
        :rtype: int
        """
        return self.numberOfStringsLessOrEqualThreeChar(s,3)-self.numberOfStringsLessOrEqualThreeChar(s,2)

        






        
